pkg/pod/entrypoint_lookup_impl_test.go

```diff
@@ -0,0 +1,172 @@
+	remotetest "github.com/tektoncd/pipeline/test"
+	"github.com/tektoncd/pipeline/test/parse"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	fakeclient "k8s.io/client-go/kubernetes/fake"
+)
+
+const (
+	username = "foo"
+	password = "bar"
+)
+
+type fakeHTTP struct {
+	reg http.Handler
+}
+
+func newfakeHTTP() fakeHTTP {
+	reg := registry.New()
+	return fakeHTTP{
+		reg: reg,
+	}
+}
+
+func (f *fakeHTTP) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Request authentication for ping request
+	// For further reference see https://docs.docker.com/registry/spec/api/#api-version-check
+	if r.URL.Path == "/v2/" && r.Method == http.MethodGet {
+		w.Header().Add("WWW-Authenticate", "basic")
+		w.WriteHeader(http.StatusUnauthorized)
+		return
+	}
+	// Check auth if we've fetching the image
+	if strings.HasPrefix(r.URL.Path, "/v2/task") && r.Method == "GET" {
+		u, p, ok := r.BasicAuth()
+		if !ok || username != u || password != p {
+			w.WriteHeader(http.StatusUnauthorized)
+			return
+		}
+	}
+	// Default to open
+	f.reg.ServeHTTP(w, r)
+}
+
+func generateSecret(host string, username string, password string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "secret",
+			Namespace: "ns",
+		},
+		Type: corev1.SecretTypeDockercfg,
+		Data: map[string][]byte{
+			corev1.DockerConfigKey: []byte(
+				fmt.Sprintf(`{%q: {"auth": %q}}`,
+					host,
+					base64.StdEncoding.EncodeToString([]byte(username+":"+password)),
+				),
+			),
+		},
+	}
+}
+
+func TestGetImageWithImagePullSecrets(t *testing.T) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	ftp := newfakeHTTP()
+	s := httptest.NewServer(&ftp)
+	defer s.Close()
+
+	u, err := url.Parse(s.URL)
+	if err != nil {
+		t.Errorf("Parsing url with an error: %v", err)
+	}
+
+	task := parse.MustParseTask(t, `metadata:name: test-create-image`)
+
+	ref, err := remotetest.CreateImageWithAnnotations(u.Host+"/task/test-create-image", remotetest.DefaultObjectAnnotationMapper, task)
+	if err != nil {
+		t.Errorf("uploading image failed unexpectedly with an error: %v", err)
+	}
+
+	imgRef, err := name.ParseReference(ref)
+	if err != nil {
+		t.Errorf("digest %s is not a valid reference: %v", ref, err)
+	}
+
+	for _, tc := range []struct {
+		name             string
+		basicSecret      *corev1.Secret
+		imagePullSecrets []corev1.LocalObjectReference
+		wantErr          bool
+	}{{
+		name:             "correct secret",
+		basicSecret:      generateSecret(u.Host, username, password),
+		imagePullSecrets: []corev1.LocalObjectReference{{Name: "secret"}},
+		wantErr:          false,
+	}, {
+		name:             "unauthorized secret",
+		basicSecret:      generateSecret(u.Host, username, "wrong password"),
+		imagePullSecrets: []corev1.LocalObjectReference{{Name: "secret"}},
+		wantErr:          true,
+	}, {
+		name:             "empty secret",
+		basicSecret:      generateSecret(u.Host, "", ""),
```

Comment:
I thought we want to test empty string here. Could you elaborate more on `explicit secret`?

Reply:
Sure! So an empty secret in my mind is something like:

```go
&corev1.Secret{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "foo",
		},
}
```

i.e. it's a secret, but it might be the wrong type, have no contents, etc.

What this is doing is creating a secret where the username and password = "", but is otherwise a valid docker config. It's roughly equivalent to a wrong username/password combination, which we already check in another test case above so there's no strong need to have a duplicate case for that here.

Reply:
I see, thank you for clarification!